test/extensions/filters/http/decompressor/decompressor_filter_test.cc

```diff
@@ -0,0 +1,281 @@
+                         ::testing::Values(true, false));
+
+TEST_P(DecompressorFilterTest, DecompressionActive) {
+  // Keep the decompressor to set expectations about it
+  auto decompressor = std::make_unique<Compression::Decompressor::MockDecompressor>();
+  auto* decompressor_ptr = decompressor.get();
+  EXPECT_CALL(*decompressor_factory_, createDecompressor())
+      .WillOnce(Return(ByMove(std::move(decompressor))));
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "mock"},
+                                                {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+
+  EXPECT_EQ(nullptr, headers_after_filter->ContentEncoding());
+
+  // FIX ME: pending decision on this.
+  EXPECT_EQ(nullptr, headers_after_filter->ContentLength());
+  EXPECT_EQ("chunked", headers_after_filter->TransferEncoding()->value().getStringView());
+  const auto* accept_encoding = headers_after_filter->get(Http::LowerCaseString{"accept-encoding"});
+  if (isRequestDirection()) {
+    EXPECT_EQ("mock", accept_encoding->value().getStringView());
+  } else {
+    EXPECT_EQ(nullptr, accept_encoding);
+  }
+
+  expectDecompression(decompressor_ptr);
+}
+
+TEST_P(DecompressorFilterTest, DecompressionActiveMultipleEncodings) {
+  // Keep the decompressor to set expectations about it
+  auto decompressor = std::make_unique<Compression::Decompressor::MockDecompressor>();
+  auto* decompressor_ptr = decompressor.get();
+  EXPECT_CALL(*decompressor_factory_, createDecompressor())
+      .WillOnce(Return(ByMove(std::move(decompressor))));
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "mock, br"},
+                                                {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+
+  EXPECT_EQ("br", headers_after_filter->ContentEncoding()->value().getStringView());
+
+  // FIX ME: pending decision on this.
+  EXPECT_EQ(nullptr, headers_after_filter->ContentLength());
+  EXPECT_EQ("chunked", headers_after_filter->TransferEncoding()->value().getStringView());
+
+  expectDecompression(decompressor_ptr);
+}
+
+TEST_P(DecompressorFilterTest, DecompressionActiveTransferEncodingPresentAlready) {
+  // Keep the decompressor to set expectations about it
+  auto decompressor = std::make_unique<Compression::Decompressor::MockDecompressor>();
+  auto* decompressor_ptr = decompressor.get();
+  EXPECT_CALL(*decompressor_factory_, createDecompressor())
+      .WillOnce(Return(ByMove(std::move(decompressor))));
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "mock, br"},
+                                                {"content-length", "256"},
+                                                {"transfer-encoding", "chunked"}};
```

Comment:
Should we set this to something custom like `mock` instead to validate that it isn't the filter setting `chunked` (like we're validating in the previous test)?

Reply:
I have actually decided to remove the transfer encoding logic. From my reading of documentation it actually doesn't seem that we should be setting it. AFAICT if the data were actually chunked we would have to transform the data chunks [per](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding):

> at the beginning of each chunk you need to add the length of the current chunk in hexadecimal format, followed by '\r\n' and then the chunk itself, followed by another '\r\n'. The terminating chunk is a regular chunk, with the exception that its length is zero. It is followed by the trailer, which consists of a (possibly empty) sequence of entity header fields. 

which this filter is not doing. Also this is a http/1.1 construct.

@rojkov am I missing something here?

Reply:
AFAIU parsing and forming data chunks gets done in the http1 codec, not in filters. It might be that the codec sets the header itself if needed, but I'm not sure. Better to checks the codec's sources.

Reply:
yep the http/1.1 codec handles this https://github.com/envoyproxy/envoy/blob/e3f1506a0c6bbe7afa0980cca58611e105f3e350/source/common/http/http1/codec_impl.cc#L139